baseplate.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/runtimebp"
+	"github.com/reddit/baseplate.go/secrets"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// Config is a general purpose config for assembling a Baseplate server
+type Config struct {
+	// Addr is the local address to run your server on.
+	//
+	// It should be in the format "${IP}:${Port}", "localhost:${Port}",
+	// or simply ":${Port}".
+	Addr string
+
+	// Timeout is the socket connection timeout for Servers that
+	// support that.
+	Timeout time.Duration
+
+	// StopTimeout is the timeout for the Stop command for the service.
+	//
+	// If this is not set, then no timeout will be set on the Stop command.
+	StopTimeout time.Duration `yaml:"stopTimeout"`
+
+	Log     log.Config
+	Metrics metricsbp.Config
+	Secrets secrets.Config
+	Sentry  log.SentryConfig
+	Tracing tracing.Config
+}
+
+// Baseplate is the general purpose object that you build a Server on.
+type Baseplate interface {
+	io.Closer
+
+	Config() Config
+	EdgeContextImpl() *edgecontext.Impl
+	Secrets() *secrets.Store
+}
+
+// Server is the primary interface for baseplate servers.
+type Server interface {
+	// Baseplate returns the Baseplate object the server is built on.
+	Baseplate() Baseplate
+
+	// Serve should start the Server on the Addr given by the Config and only
+	// return once the Server has stopped.
+	//
+	// It is recommended that you use baseplate.Serve() rather than calling Serve
+	// directly as baseplate.Serve will manage starting your service as well as
+	// shutting it down gracefully.
+	Serve() error
+
+	// Stop should stop the server gracefully and only return after the server has
+	// finished shutting down.
+	//
+	// It is recommended that you use baseplate.Serve() rather than calling Stop
+	// directly as baseplate.Serve will manage starting your service as well as
+	// shutting it down gracefully in response to a shutdown signal.
+	Stop() error
+}
+
+// Serve runs the given Server until it is given an external shutdown signal
+// using runtimebp.HandleShutdown to handle the signal and shut down the
+// server gracefully.
+//
+// If a StopTimeout is configure, Serve will wait for that duration for the
+// server to stop before timing out and returning to force a shutdown.
+//
+// This is the recommended way to run a Baseplate Server rather than calling
+// server.Start/Stop directly.
+func Serve(ctx context.Context, server Server) {
+	go runtimebp.HandleShutdown(
+		ctx,
+		func(signal os.Signal) {
+			timeout := server.Baseplate().Config().StopTimeout
+			ctx := context.Background()
+			if timeout != 0 {
+				var cancel context.CancelFunc
+				ctx, cancel = context.WithTimeout(ctx, timeout)
+				defer cancel()
+			}
+
+			var err error
+			ch := make(chan error)
+			go func() {
+				ch <- server.Stop()
+			}()
+
+			// Wait for either the context to timeout or Stop() to finish.
+			select {
+			case <-ctx.Done():
+				err = ctx.Err()
+			case e := <-ch:
+				err = e
+			}
```

Comment:
Originally this was handled by `Server.Stop` (largely because `http.Server` handles this), but it seemed like it made more sense to just do this here rather than in each implementation.